contracts/ProcedureTable.sol

```diff
@@ -498,39 +839,67 @@ library ProcedureTable {
     }
 
     function remove(Self storage /* self */, bytes24 key) internal returns (bool success) {
-        Procedure memory p1 = _getProcedureByKey(uint192(key));
+        // Procedure memory p1 = _getProcedureByKey(uint192(key));
 
-        if (p1.keyIndex == 0)
+        // Get the key index of the procedure we are trying to delete.
+        uint192 p1Index = _getProcedureIndex(uint192(key));
+
+        // If the value does not exist in the procedure list we return 0.
+        if (p1Index == 0)
             return false;
         // When we remove a procedure, we want to move another procedure into
         // this "slot" to keep the active keys contiguous
         // First we get the storage address of the Procedure Table Length
-        uint248 lenP = _getLengthPointer();
+        uint256 lenP = _getLengthPointer();
         // We then dereference that to get the value
-        uint256 len = _get(0, lenP);
-
-        if (p1.keyIndex <= len) {
-            // Move an existing element into the vacated key slot.
-            uint248 p1P = _getProcedurePointerByKey(uint192(key));
-            uint248 key1P = _getKeyPointerByIndex(p1.keyIndex);
+        uint256 len = _get(lenP);
+
+        if (p1Index <= len) {
+            // Three things need to happen:
+            //   When a procedure is deleted from the kernel:
+            // 1. If the procedure key is the same as the Entry Procedure Key, abort and
+            // throw an error.
+            // 2. If the procedure key does not exist in the list (i.e. when looking on the
+            // procedure heap no procedure index is associated with it), abort and throw
+            // an error.
+            // 3. The length value is decremented by one.
+            // 4. If the procedure being deleted is not the last in the list (i.e. it’s procedure
+            // index does not equal the length of the procedure list), the last in the list is
+            // copied to overwrite the key being deleted. This also accounts for the case
+            // of an empty list.
 
```

Comment:
```suggestion
        // 5. The capability list of the procedure is iterated per each cap type and cleared from storage.
```

Reply:
The current `#remove` implementation doesn't clear the c-list of the procedure.

Reply:
Yep, It's currently specified as optional in the spec. What do you think? We've got three options:

1. Do nothing.
2. Set all the c-list lengths to zero.
3. Clear everything.

Setting the c-lists to zero length may have some security benefits if we re-register the same key, making sure it doesn't get any zombie capabilities. Let's update the spec with whatever we decide first.

Reply:
I'm torn between 2 and 3. 2 would make upgrades much much cheaper of course.

For the spec I'm guessing we could say that at the end of execution, all deleted Procedures should have their c-list cleared. That would require a designated place in kernel storage for procedureId's to be cleared. 

For an alpha I guess we could just do whichever is easier.

Reply:
I've been thinking along 2 lines for this:

1. The code that prevents reuse of old caps should be simple and solid. If we rely on two different processes to work together that is more complicated
2. Deletion should be cheap and simple, and any expensive operations should be borne by the registration process rather than the deletion process.

Based on these I liked 1, it's kind of the opposite of security though obscurity.

Having said that, there's something nice about clearing the capabilities away. I don't think there's a clear solution here, just a bunch of small trade offs.

I also like 2, it has an elegance, but 3 is the simplest specification wise, which is also something to consider.

You want to pick?

Reply:
Moved to #129.